src/radical/pilot/agent/scheduler/base.py

```diff
@@ -310,14 +314,33 @@ def create(cls, cfg, session):
     #
     # NOTE: any scheduler implementation which uses a different nodelist
     #       structure MUST overload this method.
+
+    # MING: This function is used to update the node list by taking in a list
+    #       of slots that have been assigned an update them to the new state.
+    #       The nodes are represented as a dictionary, where as the slots are
+    #       represented as a dictionary of a different structure. This confused
+    #       me at first when I tried to read it.
+
     def _change_slot_states(self, slots, new_state):
 
         for node_name, node_uid, cores, gpus in slots['nodes']:
 
             # Find the entry in the the slots list
+
+            # MING: Assuming 'uid' is the ID of the node, it seems a bit
+            #       wasteful to have to look at all of the nodes available
+            #       for use if at most one node can have that uid. Maybe
+            #       it would be worthwhile to simply keep a list of nodes
+            #       that we would read, and keep a dictionary that maps
+            #       the uid of the node to the location on the list?
```

Comment:
So, basically creating an uid-lookup index into the list? Yeah, makes sense as an optimization, since that index would be static over the lifetime of the agent.  Would you be up to implementing this?  Otherwise we could put it up for a task assignment, its useful, small and fairly self-contained...  Nice one!

Reply:
I don't mind implementing this!